### ISSUE TITLE
Auth0 Electron Callback URL mismatch

### ISSUE THREAD

**Issue by Toosick** [auth0/lock#1217]
@chenkie @auth0 I can't seem to get `file:///` to work with our Electron app. I was able to get things working by adding the full callback `file:///` url into the `Allowed Callback URLs` setting. The callbacks work just fine on the web using the callback with `https://*.netlify.com/login`. Does it now have to be something like `file://*` or `file:///*`?

Client ID: QR579kSOEqhOqPAfqjfTPHaz5hJPvEKT
Auth0 lock version: 11.0.1

### Callback mismatch
![Callback mismatch](https://image.prntscr.com/image/oWZ5OXuXQxaVU20Ugck_yQ.png "Callback mismatch")

### Auth0 lock setup
![Auth0 lock setup](https://image.prntscr.com/image/up0m50tWRxyuLXOfTaC00w.png "Auth0 lock setup")

**Comment by luisrudge** [comment]
Hi, did you follow https://auth0.com/docs/quickstart/native/electron/01-login ?

**Comment by Toosick** [comment]
@luisrudge Yes, as you can see I already added the callbacks from that article.

**Comment by Toosick** [comment]
@luisrudge Any ideas?

**Comment by darkyen** [comment]
I looked into the issue, unfortunately our electron quickstart is a little dated, we recently added support for custom apis that allow better integration for native applications and clis, desktops and native apps should either be using the platform specific authentication / authorization layer (for example WebAuthenticationBroker in Windows) and use PKCE Grant. Which is recommended by OIDC.
 
In a short term you can get around this by passing 
```
{
   redirectUri: 'https://<yourdomain>.auth0.com/mobile'
}
```

However you'll need to open a new window and listen for the changes on URL and handle it accordingly.  A much better flow would be to use [auth0-pkce](https://auth0.com/docs/api-auth/tutorials/authorization-code-grant-pkce)

**Comment by bkniffler** [comment]
Could you guys provide an up-to-date example for the pkce flow / electron? That would be supercool!

**Comment by Toosick** [comment]
@bkniffler Are you getting the same errors I'm geting?

**Comment by bkniffler** [comment]
I managed to get it working @Toosick . You can checkout the result here (for electron-renderer): https://github.com/olymp/olymp/blob/master/packages/olymp-auth/auth0-electron.es6

Also you will need this handler in your electron main process:
https://github.com/olymp/olymp-webpack/blob/master/packages/olymp-webpack-electron/main.es6#L124

Don't forget to put `com://callback` into the auth0 login callback whitelist. You can also use it as a logout callback.

The way it works is, it will 
- open a new window in the renderer via window.open with a href pointing to auth0 authorize endpoint
- hook into this window through electron-main process (via 'new-window' event) and wait until redirection to callback happens (via 'will-navigate' and 'did-get-redirect-request' events)
- communicate the resulting code from electron-main to electron-renderer through ipc
- exchange the code for refresh and access tokens, store them in electron localStorage
- whenever expired, get a new access token

Shouldn't be too hard to abstract the essential stuff out into some npm package.

btw I found that not using auth0-js/auth0-lock is way more easy, way less limiting and way lighter on your bundle size.

I'd love to know if this approach is correct @luisrudge @darkyen

**Comment by adeperio** [comment]
Hi all, looking to implement this too. Are we saying that the Auth0 sdks need to be updated for PKCE flow in native electron apps at the moment? We should implement it ourselves as @bkniffler has?

**Comment by adeperio** [comment]
@darkyen Ahh great thanks for the update! I've actually implemented the PKCE flow similar to @bkniffler, by:

1. Using BrowserWindow.loadUrl to get to the Hosted Login page on the /authorize endpoint (and sending through a challenge) 
2. And then, after successfully authenticating with passwordless, using the 'did-get-redirect-request' event on the BrowserWindow to capture the redirect and to fire off a request (with the verifier and the received auth code) for the access token on the /token endpoint. 

This seems to be working ok. Is it a hard requirement to use the system's native browser for this flow in Electron?

**Comment by adeperio** [comment]
Hi all, for reference I've created a gist as well of what I have worked on so far on this. It's using the BrowserWindow renderer to execute the PKCE flow (up to when it gets the access code, still working on saving it to local storage etc)

https://gist.github.com/adeperio/73ce6680d4b80b45e624ab62bacfbdca

I hope that's helpful too? Do you think this looks like it's on the right track as well?

**Comment by bkniffler** [comment]
Any news @darkyen? I still find auth0 universal auth difficult to implement for apps that should work on electronjs, web (+ios standalone), cordova with a shared codebase. And there is so much outdated docs out there.

**Comment by KaitoKid** [comment]
@darkyen Any update on this?

Also, @adeperio thanks for your gist, but it seems to break upon DUO MFA with 'Duo is not defined'

**Comment by bkniffler** [comment]
Man, its really frustrating. Auth0 is such a great platform, I'd just wish their js tools and docs would be as great as firebase. Its taken me 15 minutes to implement firebase with refresh tokens etc in my electron app and the react-native app.

**Comment by adeperio** [comment]
Similar experience here @bkniffler, at least for Electron Support, it's a little bit like detective work at the moment (and a lot of time just implementing the API flows by hand out of the Auth0 docs). @KaitoKid I haven't tested that gist with MFA unfortunately. It's meant to only be as an example code for basic PKCE flow right now.

**Comment by uotw** [comment]
@darkyen Any official auth0 electron pkce auth quickstart coming down the pike soon?  My app will be breaking in 2 weeks using the deprecated lock quickstart code unless I fix this issue in an update.

**Comment by MacroChip** [comment]
I have been trying to show the Universal Login hosted page in a BrowserWindow that shares a partition with the rest of my app. My assumption is that the hosted page sets a cookie that the rest of my app would use since the hosted page shares a partition with the rest of my app. I cannot get it working. Is there anything wrong with this approach?

**Comment by bkniffler** [comment]
In electron you'll have to use accessToken/refreshToken and manually store/read them to some local store like localStorage instead of using cookies, which wouldn't work.

**Comment by KaitoKid** [comment]
You can try storing to cookies and using ipc from the front end or sessionStorage (which is how I got it to work)

**Comment by uotw** [comment]
@MacroChip I was able to get it working with @adeperio 's gist above.
https://gist.github.com/adeperio/73ce6680d4b80b45e624ab62bacfbdca

**Comment by adeperio** [comment]
@MacroChip I would try and look at the node-keytar npm library for securely storing the access token and especially the refresh token (this isn't in my gist). On mac this uses the Keychain, and on windows this uses the Credential Vault. All of this happens under the hood with node-keytar, and you can just interact with it through its API

**Comment by uotw** [comment]
Forgive me if this is completely ignorant, but what's the security risk if the electron user has access to this token and refresh key?  Anything he does with it only affects his auth0 info and registration.  You could just destroy localfilestorage on exit of the app.

**Comment by adeperio** [comment]
Hey @uotw it's less about the user having access to it (they're supposed to) it's more if something else gains access to it. The access token has some protection with whatever expiry that you set on it, but the refresh token is long lived, and sits on the client, so that should be protected where possible. Lots of other companies use the keychain to store these credentials too:

![image](https://user-images.githubusercontent.com/1326430/43056042-2ecaaffc-8e7d-11e8-9d94-73f660239659.png)

Also, I found the docs on Auth0 around access tokens and refresh tokens really useful reading too. 
https://auth0.com/docs/tokens/refresh-token/current
https://auth0.com/docs/tokens/access-token

**Comment by jeaye** [comment]
So, does auth0-lock just simply not work with Electron? Do we need to manually create a window and go through the flow shown in adeperio's gist?

I've also tried auth0-js, but ran into similar callback issues. Will someone from Auth0 please confirm what the best approach is for Electron users and provide some official documentation?

**Comment by luisrudge** [comment]
@jeaye Hi! Sorry about the trouble. We deprecated that sample. We're working on a new one with the recommended approach.

**Comment by bkniffler** [comment]
Any timeframe for these guidances? Months? Weeks?

**Comment by jeaye** [comment]
If we're closing this one since Electron isn't supported, here's a ticket for supporting Electron: https://github.com/auth0/lock/issues/1467 Let's discuss more there.

**Comment by bkniffler** [comment]
It has been a year now and still no official guide on how to use universal login with electron. There is a few unofficial examples like https://github.com/uotw/auth0-PKCE-for-electron-quickstart and an npm package https://www.npmjs.com/package/electron-auth0-login. 

But they have custom logic for url building and crypto (https://github.com/uotw/auth0-PKCE-for-electron-quickstart/blob/master/js/AuthService.js), which I think would be better housed in an official auth0-login library that works for any JS target (web, electron, react-native, cordova).

**Comment by dallasclymer** [comment]
> It has been a year now and still no official guide on how to use universal login with electron. There is a few unofficial examples like https://github.com/uotw/auth0-PKCE-for-electron-quickstart and an npm package https://www.npmjs.com/package/electron-auth0-login.
> 
> But they have custom logic for url building and crypto (https://github.com/uotw/auth0-PKCE-for-electron-quickstart/blob/master/js/AuthService.js), which I think would be better housed in an official auth0-login library that works for any JS target (web, electron, react-native, cordova).
> 
> edit: electron-auth0-login does actually work quite well

Does this solve the callback issues for file:// instead of a URL.

**Comment by bkniffler** [comment]
@dallasclymer The callback issue is being worked around by opening a local express server within the main process to serve the renderer.. so yeah, it works somewhat.
